qcodes/utils/ndim_analysis/dataset/doNd.py

```diff
@@ -0,0 +1,309 @@
+AxesTupleList = Tuple[List[matplotlib.axes.Axes],
+                      List[Optional[matplotlib.colorbar.Colorbar]]]
+AxesTupleListWithRunId = Tuple[int, List[matplotlib.axes.Axes],
+                      List[Optional[matplotlib.colorbar.Colorbar]]]
+
+
+def _process_params_meas(param_meas: ParamMeasT) -> List[res_type]:
+    output = []
+    for parameter in param_meas:
+        if isinstance(parameter, _BaseParameter):
+            output.append((parameter, parameter.get()))
+        elif callable(parameter):
+            parameter()
+    return output
+
+
+def _register_parameters(
+        meas: Measurement,
+        param_meas: List[ParamMeasT],
+        setpoints: Optional[List[_BaseParameter]] = None
+) -> None:
+    for parameter in param_meas:
+        if isinstance(parameter, _BaseParameter):
+            meas.register_parameter(parameter,
+                                    setpoints=setpoints)
+
+
+def _register_actions(
+        meas: Measurement,
+        enter_actions: ActionsT,
+        exit_actions: ActionsT
+) -> None:
+    for action in enter_actions:
+        # this omits the possibility of passing
+        # argument to enter and exit actions.
+        # Do we want that?
+        meas.add_before_run(action, ())
+    for action in exit_actions:
+        meas.add_after_run(action, ())
+
+
+
+def _set_write_period(
+        meas: Measurement,
+        write_period: Optional[float] = None
+) -> None:
+    if write_period is not None:
+        meas.write_period = write_period
+
+
+@contextmanager
+def _catch_keyboard_interrupts() -> Iterator[Callable[[], bool]]:
+    interrupted = False
+    def has_been_interrupted():
+        nonlocal interrupted
+        return interrupted
+    try:
+        yield has_been_interrupted
+    except KeyboardInterrupt:
+        interrupted = True
+
+
+
+def do0d(
+    *param_meas:  ParamMeasT,
+    write_period: Optional[float] = None,
+    do_plot: bool = True
+) -> AxesTupleListWithRunId:
+    """
+    Perform a measurement of a single parameter. This is probably most
+    useful for an ArrayParamter that already returns an array of data points
+
+    Args:
+        *param_meas: Parameter(s) to measure at each step or functions that
+          will be called at each step. The function should take no arguments.
+          The parameters and functions are called in the order they are
+          supplied.
+        do_plot: should png and pdf versions of the images be saved after the
+            run.
+
+    Returns:
+        The run_id of the DataSet created
+    """
+    meas = Measurement()
+    _register_parameters(meas, param_meas)
+    _set_write_period(meas, write_period)
+
+    with meas.run() as datasaver:
+        datasaver.add_result(*_process_params_meas(param_meas))
+
+    return _handle_plotting(datasaver, do_plot)
+
+
+
+
+
+def do1d(
+    param_set: _BaseParameter, start: float, stop: float,
+    num_points: int, delay: float,
+    *param_meas: ParamMeasT,
```

Comment:
Is it possible to pass in a function that can break the measurement? I might want to just look at the result of the latest measurement and if it passes some threshold break the measurement. Creating the function to do the check can easily be done outside the doNd functions but what should the flag/function return be that correctly stops the measurement and exists with save of dataset and figures?